### PR TITLE
Annotation sets functionalities - WIP

### PR DESCRIPTION
**Pull request by drewlr**
Polishing and enabling annotations management functionalities

### CODE REVIEW

File: remo/annotation_utils.py [hunk 1]
```diff
@@ -65,7 +65,9 @@ def create_tempfile(annotations: List[Annotation]) -> str:
     
     annotation_task = annotations[0].task
     prepared_data = prepare_annotations_for_upload(annotations, annotation_task)
-    list_of_classes = [i_ele[1] for i_ele in prepared_data]
+    
+    # getting a list of classes. Skipping the first row as it contains the csv header
+    list_of_classes = [i_ele[1] for i_ele in prepared_data[1:]]
```
Comment (vmanilo) [inline remo/annotation_utils.py]:
```suggestion
    classes = [row[1] for row in data[1:]]
```

File: remo/domain/dataset.py [hunk 2]
```diff
@@ -120,25 +120,45 @@ def annotation_sets(self) -> List[AnnotationSet]:
...
         annotation_set = self.get_annotation_set(annotation_set_id)
         temp_path, list_of_classes = create_tempfile(annotations)
         
-        if not annotation_set:
-
-            self.create_annotation_set(annotation_task=annotations[0].task, name='my_ann_set',
-                                       classes = list_of_classes, path_to_annotation_file = temp_path)
         
-        elif annotation_set.task != annotations[0].task:
+        if (not annotation_set) or (annotation_set.task != annotations[0].task) or (annotation_set_id==-1):
```
Comment (vmanilo) [inline remo/domain/dataset.py]: when `annotation_set_id == -1`?

File: remo/domain/dataset.py [hunk 3]
```diff
@@ -271,12 +291,13 @@ def get_annotation_set(self, annotation_set_id: int = None) -> AnnotationSet:
...
         if not annotation_set_id:
             return self.default_annotation_set()
-
-        annotation_set = self.sdk.get_annotation_set(annotation_set_id)
-        if annotation_set and annotation_set.dataset_id == self.id:
-            return annotation_set
-        else:
-            raise Exception('Annotation set with ID = {} is not part of dataset {}. You can check the list of annotation sets in your dataset using dataset.annotation_sets()'.format(annotation_set_id, self.name))
+        
+        if annotation_set_id != -1:
```
Comment (vmanilo) [inline remo/domain/dataset.py]: when `annotation_set_id is -1`?

Comment (drewlr) [reply remo/domain/dataset.py]: @vmanilo I think is bad design on my side. Need to find a way to tell the function to create a new annotation set

File: remo/domain/dataset.py (hunk 2 above)
Comment (drewlr) [reply remo/domain/dataset.py]: see below

Comment (vmanilo) [reply remo/domain/dataset.py]: `(optional) annotation_set_id: annotation set id. If annotation_set_id = -1, a new annotation set will be created`

I don't think it's needed, `annotation_set_id` can be just None

if you want to force creating annotation set, I guess would be better to add boolean flag like `create_new`

File: remo/api.py [hunk 4]
```diff
@@ -198,13 +198,19 @@ def upload_files(
             annotation_set_id=annotation_set_id,
         )
         r = self.post(url, files=files, data=data)
-
+        json_resp = r.json()
+        
+        #TODO ALR: there can be more elements in list jsoned_resp['errors']?
+        if (r.status > 400)  and ('errors' in json_resp):
+            raise Exception('Error description:' + json_resp['errors'][0])
```
Comment (vmanilo) [inline remo/api.py]:
```suggestion
            raise Exception('Error description:' + '\n'.join(json_resp['errors']))
```

File: remo/domain/annotation.py [hunk 5]
```diff
@@ -34,7 +34,15 @@ def __init__(self, img_filename: str = None, classes=None, object=None):
         self.classes = classes if isinstance(classes, list) else [classes]
         self.object = object
         self.coordinates = None
+        self.type = None
+        
+    def __str__(self):
+        my_representation =  "Annotation: {classes} (type:{ann_type}, file:{filename})".format(classes=self.classes, ann_type=self.type, filename=self.img_filename) 
+        
+        return my_representation
```
Comment (vmanilo) [inline remo/domain/annotation.py]:
```suggestion
        return "Annotation: {classes} (type:{ann_type}, file:{filename})".format(classes=self.classes, ann_type=self.type, filename=self.img_filename) 
```

File: remo/domain/annotation.py [hunk 6]
```diff
@@ -55,6 +63,7 @@ def bbox(self, values: List[int]):
 
         self.object = Annotation.Bbox(*values)
         self.coordinates = values
+        self.type = 'Bounding Box'
```
Comment (vmanilo) [inline remo/domain/annotation.py]: I think `type` relates to Annotation.Bbox class, and should be moved there

similar for Annotation.Segment

File: remo/domain/dataset.py [hunk 7]
```diff
@@ -111,34 +123,56 @@ def fetch(self):
...
-        if not annotation_set:
-
-            self.create_annotation_set(annotation_task=annotations[0].task, name='my_ann_set',
-                                       classes = list_of_classes, path_to_annotation_file = temp_path)
-        
-        elif annotation_set.task != annotations[0].task:
+        if annotation_set and create_new_annotation_set:
+            raise Exception("You passed an annotation set but also set create_new_annotation_set = True. You can't have both.")
+            
+        if (not annotation_set) or (annotation_set.task != annotations[0].task) or (create_new_annotation_set==True):
```
Comment (vmanilo) [inline remo/domain/dataset.py]:
```suggestion
        if create_new_annotation_set or (not annotation_set) or (annotation_set.task != annotations[0].task):
```

File: remo/domain/annotation.py (hunk 5 above)
Comment (drewlr) [reply remo/domain/annotation.py]: I stick by the more verbose form! :)

File: remo/domain/annotation.py (hunk 6 above)
Comment (drewlr) [reply remo/domain/annotation.py]: @vmanilo it's just that we don't have a sub class for image classification at the moment. Also not clear how to handle tags from code yet - I would leave this for a later smaller refactoring, do you agree?

Comment (vmanilo) [reply remo/domain/annotation.py]: ok

**Pull request merged**